### PR TITLE
Try out test helper

### PR DESCRIPTION
**Pull request by matthijsgroen**
## What

Experiment with the `setComponent` concept that is also used @ Payt. 

☝️ @nickve28 

## Why

This will prove this model, and work as reference for other codebases. This can then also be referenced from the FE Golden Path documentation

## Code Review

Please consider the following checklist when reviewing this Pull Request.  
More background and details [here](https://github.com/kabisa/kabisa-guide/blob/master/organization/process/code-review/README.md).

* [ ] Does the code **actually solve** the problem it was meant to solve?
* [ ] Is the code covered by **unit tests**? **Integration tests**?
* [ ] Does anything here need **documentation**? (Focus on *why*, not *what.*)
* [ ] Does any of this code deal with **privacy sensitive information** or affects **security**? Ask an additional reviewer.
* [ ] Is the code easy to **understand** and **change** in the future?
* [ ] Is the same code or concept **duplicated**? Find a balance between DRYness and readability.
* [ ] Does the code reasonably adhere to the Kabisa [**coding standards**](https://github.com/kabisa/kabisa-guide/blob/master/organization/process/code-review/README.md)?
* [ ] Be kind.

### CODE REVIEW

**Comment by nickve28** [comment]
Nice! Erg mooie uitvoer met TS typings ook. Wellicht zou wat JSdoc toevoegen ook helpen, dan heb je 2 vliegen in 1 klap (ook voor het golden path meteen)

**Comment by nickve28** [comment]
Je zou er nog voor kunnen kiezen om een 'wait for element to be removed' / loading indicator afwachten stap toe te voegen, dat heeft payt ook, ipv de standaard

```javascript
const view = renderComponent()
await waitForElementTobeRemoved(() => expect(screen.queryByText("loading...")).not.toBeIntheDocument()
```

```
  setComponent(MyComponent, { waitForElementToBeRemoved: () => screen.queryByTestId('loading-icon') })
  ```
  
  zij doen het zo, met een 'await' op renderComponent bijv

File: src/support/testing/testComponent.tsx [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+export type Decorator<
+  TName extends string,
+  TSettings extends Record<string, unknown>,
+> = {
+  name: TName;
+  settings: TSettings;
+  decorator: (Component: React.FC, settings: TSettings) => JSX.Element;
+};
+
+export const setComponent = <
```
[the remark this answers is not in the document]
Comment (matthijsgroen) [reply src/support/testing/testComponent.tsx]: Been reduced, but would complicate setup a lot if trying to make it lower

File: src/support/testing/testDecorators.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import { Decorator } from "./testComponent";
+import { MockedProvider } from "@apollo/client/testing";
+import { MemoryRouter, Route } from "react-router-dom";
+import { ApolloCache } from "@apollo/client/cache";
+import { Context as ResponsiveContext } from "react-responsive";
+
+export const responsiveDecorator: Decorator<"responsive", { width: number }> = {
+  name: "responsive",
+  settings: { width: 1200 },
+  decorator: (Component, settings) => (
```
[the remark this answers is not in the document]
Comment (matthijsgroen) [reply src/support/testing/testDecorators.tsx]: We are unable to set a displayName in this construct

File: src/support/testing/testDecorators.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+export const dataDecorator = <TSerialized extends object>(
+  mocks?: any,
+  cache?: ApolloCache<TSerialized>,
+): Decorator<
+  "application",
+  { useTypeName: boolean; mocks?: any; cache?: ApolloCache<TSerialized> }
+> => ({
+  name: "application",
+  settings: { useTypeName: true, mocks, cache },
+  decorator: (Component, settings) => (
```
[the remark this answers is not in the document]
Comment (matthijsgroen) [reply src/support/testing/testDecorators.tsx]: We are unable to set a displayName in this construct

File: src/support/testing/testDecorators.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+export const routingDecorator = (): Decorator<
+  "routing",
+  { paths: Record<string, React.ReactNode> }
+> => ({
+  name: "routing",
+  settings: {
+    paths: {},
+  },
+  decorator: (Component, settings) => (
```
[the remark this answers is not in the document]
Comment (matthijsgroen) [reply src/support/testing/testDecorators.tsx]: We are unable to set a displayName in this construct

File: src/support/testing/testDecorators.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
+import { Decorator } from "./testComponent";
+import { MockedProvider } from "@apollo/client/testing";
+import { MemoryRouter, Route } from "react-router-dom";
+import { ApolloCache } from "@apollo/client/cache";
+import { Context as ResponsiveContext } from "react-responsive";
+
+export const responsiveDecorator: Decorator<"responsive", { width: number }> = {
+  name: "responsive",
+  settings: { width: 1200 },
+  Decorator: ({ Component, settings }) => (
```
[the remark this answers is not in the document]
Comment (matthijsgroen) [reply src/support/testing/testDecorators.tsx:10]: We are unable to set a displayName in this construct

File: src/support/testing/testDecorators.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+export const dataDecorator = <TSerialized extends object>(
+  mocks?: any,
+  cache?: ApolloCache<TSerialized>,
+): Decorator<
+  "application",
+  { useTypeName: boolean; mocks?: any; cache?: ApolloCache<TSerialized> }
+> => ({
+  name: "application",
+  settings: { useTypeName: true, mocks, cache },
+  Decorator: ({ Component, settings }) => (
```
[the remark this answers is not in the document]
Comment (matthijsgroen) [reply src/support/testing/testDecorators.tsx:26]: We are unable to set a displayName in this construct

File: src/support/testing/testDecorators.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+export const routingDecorator = (): Decorator<
+  "routing",
+  { paths: Record<string, React.ReactNode> }
+> => ({
+  name: "routing",
+  settings: {
+    paths: {},
+  },
+  Decorator: ({ Component, settings }) => (
```
[the remark this answers is not in the document]
Comment (matthijsgroen) [reply src/support/testing/testDecorators.tsx:45]: We are unable to set a displayName in this construct

File: src/support/testing/testDecorators.tsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        </Route>
+      ))}
+    </MemoryRouter>
+  ),
+});
+
+export const tableDecorator: Decorator<"table", Record<string, unknown>> = {
+  name: "table",
+  settings: {},
+  Decorator: ({ Component }) => (
```
[the remark this answers is not in the document]
Comment (matthijsgroen) [reply src/support/testing/testDecorators.tsx:60]: We are unable to set a displayName in this construct

File: src/support/testing/testComponent.tsx [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+  /**
+   * Name of the decorator, can be used in `updateDecorator` to update the settings
+   */
+  name: TName;
+  /**
+   * Initial settings of this decorator, van be updated through `updateDecorator`
+   */
+  settings: TSettings;
+  /**
+   * Function do decorate incoming `Component`. This can be part of a larger chain.
```
Comment (nickve28) [inline src/support/testing/testComponent.tsx]: 'to decorate' denk ik

**Review by nickve28 — COMMENTED** [review]
lijkt mij prima zo

**Review by nickve28 — APPROVED** [review]

**Pull request merged**